azkaban-exec-server/src/test/java/azkaban/execapp/FlowRunnerTestBase.java

```diff
@@ -170,4 +171,8 @@ protected void succeedJobs(final String... jobs) {
     }
   }
 
+  protected void assertEvents(final Type... types) {
```

Comment:
Navigation is not a big issue although I would prefer minimizing the redirection when possible. e.g.   protected void assertEvents(final Type... types) {
    eventCollector.assertEvents(types);
  }
is one of such redirection. Now I have to remember where the eventCollector member comes from and where it is modified. I am looking at the base class. But the member is modified in the subclasses.

The bigger issue is to reduce the mental overhead of figuring out the states.  I prefer limiting the scope of a variable as much as possible. e.g. What's the advantage for turning the eventCollector from a local variable into a member variable? as I asked earlier.
 I would prefer composition over inheritance. e.g. it looks like the FlowRunnerTestBase can be turned into a utility class with static methods and pass in the runner object when needed to the methods that need it. 

See my pull request #1151 for an example. the new prepareProject method is relatively simple to understand in terms of what are the inputs and outputs. 

Reference:
https://stackoverflow.com/questions/11343840/favor-composition-over-inheritance
And the "Effective Java" book item 14: favor-composition-over-inheritance

If you can break up this pull request into simpler multiple ones, we can merge the first commit while we discuss the rest. This is another advantage of smaller and independent pull requests i.e. we can treat different parts differently.